ygen/enumgen.go

```diff
@@ -229,47 +247,40 @@ func (s *enumSet) identityBaseKey(i *yang.Identity) string {
 // enumeration). It also returns the type name component of the identification
 // key for use in the name generation, as well as the enumeration type object,
 // if needed.
-func (s *enumSet) enumeratedTypedefKey(e *yang.Entry, noUnderscores, useDefiningModuleForTypedefEnumNames bool) (string, string, *yang.YangType, error) {
-	enumType := e.Type
-	typeName := e.Type.Name
+func (s *enumSet) enumeratedTypedefKey(args resolveTypeArgs, noUnderscores, useDefiningModuleForTypedefEnumNames bool) (string, string, error) {
+	typeName := args.yangType.Name
+	definingType, err := util.DefiningType(args.yangType, args.contextEntry.Type)
+	if err != nil {
+		return "", "", err
+	}
 
 	// Handle the case whereby we have been handed an enumeration that is within a
 	// union. We need to synthesise the name of the type here such that it is based on
 	// type name, plus the fact that it is an enumeration.
-	if e.Type.Kind == yang.Yunion {
-		enumTypes := util.EnumeratedUnionTypes(e.Type.Type)
-
-		switch len(enumTypes) {
-		case 1:
-			// Under a union, the correct enumType reference is the enumeration subtype.
-			enumType = enumTypes[0]
-			// We specifically say that this is an enumeration within the leaf.
-			if noUnderscores {
-				typeName = fmt.Sprintf("%sEnum", enumTypes[0].Name)
-			} else {
-				typeName = fmt.Sprintf("%s_Enum", enumTypes[0].Name)
-			}
-		case 0:
-			return "", "", nil, fmt.Errorf("enumerated type had an empty union within it, path: %v, type: %v, enumerated: %v", e.Path(), e.Type, enumTypes)
-		default:
-			return "", "", nil, fmt.Errorf("multiple enumerated types within a single enumeration not supported, path: %v, type: %v, enumerated: %v", e.Path(), e.Type, enumTypes)
+	if definingType.Kind == yang.Yunion {
+		// We specifically say that this is an enumeration within the leaf.
+		if noUnderscores {
+			typeName = fmt.Sprintf("%s%s", definingType.Name, enumeratedUnionSuffix)
+		} else {
+			typeName = fmt.Sprintf("%s_%s", definingType.Name, enumeratedUnionSuffix)
```

Comment:
Does the old behaviour still work the same with `useDefiningModuleForTypedefEnumNames` set to false?

Reply:
Yes. When `useDefiningModuleForTypedefEnumNames` is false, `enumeratedTypedefKey()` is only ever called with the entry being a typedef, meaning the `DefiningType` of the enum entry is always itself, producing the same behaviour.

As for the `len(enumTypes)` part of the old code, this new code makes the decision that *multiple enums within a union is now allowed*. This is a deviation from before, but it's a deviation that's in the right direction and backwards-compatible.

Reply:
There is a minor backwards-compatibility issue with when to append `_Enum` that's now fixed.